readme.md
# @brixtol/rollup-utils

Various rollup plugin utilities for working with packages in the [Brixtol Textiles](https://brixtoltextiles.com) monorepo workspace. These utilities are a collection of commonly used modules leveraged in development for both closed and open source and modules across the Brixtol monorepo.

> This module is also provided within [@brixtol/rollup-config](https://github.com/BRIXTOL/rollup-config). If you are using that module then you do not need to install this.

### Why?

We here at Brixtol Textiles leverage [Rollup](https://rollupjs.org/guide/en/) exclusively. This package is a collection of various utilities that assist in the bundling and transpilation process of JavaScript/TypeScript packages. The utilities exported in this module provide a clean and easy way to execute commonly used functions. Using this enables us to provide a DRY configuration export and improves productivity in the workspace.

# Utilities

| Export                       | Description                                                                |
| ---------------------------- | -------------------------------------------------------------------------- |
| `path(string)`               | Resolves a Path to location to current working directory.                  |
| `env.*`                      | Provides ENV operations for `dev`, `prod` and `watch` builds.              |
| `config.*`                   | Enables `package.json` or other config files within `rollup.config.mjs`    |
| `banner({}, string)`         | Generates comment banner in bundles that includes License and information  |
| `jsonmin(string)`            | Minifies JSON, used with [rollup-plugin-copy](https://git.io/J0Lv9)        |
| `date(utc?: Date \| string)` | Returns a pretty formatted date with number suffix, eg: `1st October 2021` |

# Install

```cli
pnpm add @brixtol/rollup-utils --save-dev
```

> This module is used together with [@brixtol/rollup-config](https://github.com/brixtol/rollup-config) with all exports exposed from that module.

### Usage

Typings explain the utilities in good detail. It's important that when using the `env` util environment values are passed on the command line using the Rollup `--environment` flag. Pass `prod` for production and either omit or pass `dev` for development.

> Custom flags are not supported, only `prod` and `dev` are available, you can however pass a flag who's value evaluates to a `'true'` or `'false'` boolean.

<!-- prettier-ignore -->
```js
import { env, config } from "@brixtol/rollup-utils";

export default {
  input: "src/file.js",
  output: {
    format: 'cjs',
    file: config.output.cjs, // The exports.require value in package.json
    sourcemap: env.is('dev', 'inline') // Inline sourcemap in development else false
    interop: 'default'
  },
  plugins: env.if('div')(
    [
      // Development plugins
    ]
  )(
    [
      // Production plugins
    ]
  )
}
```

## `env.*`

The `env` export exposes a bunch of utility methods the pertain to the environment execution you a running a rollup build. This export is used together with rollup `--environment` flags.

#### `env.dev`

Returns a `boolean` of either `true` or `false` depending on the `--environment` flag value

#### `env.prod`

Returns a `boolean` of either `true` or `false` depending on the `--environment` flag value

#### `env.watch`

Returns a `boolean` of either `true` or `false` depending on the `--environment` flag value

#### `env.vars`

The `env.vars` method will return a parsed list of environment variables contained within a `.env` file. This method uses the [dotenv](https://www.npmjs.com/package/dotenv) package for this.

> The `.env` file will be parsed using the default options of [dotenv](https://www.npmjs.com/package/dotenv).

#### `env.if`

The `env.if()` allows us to use single file for development and production bundles. When an `--environment` flag is passed with a of value of `prod` the plugins are concatenated, so first curried parameter is combined with the second curried parameter, which should both be an array list of plugins[]. If you pass a `boolean` type to the first argument in the curry, then `false` will load `dev` plugins whereas `true` executes a concatenation.

The `dev` is default, so running `rollup -c -w` results in:

<!-- prettier-ignore -->
```ts
import { env } from '@brixtol/rollup-utils'

env.if('dev')([ plugin.commonjs(), plugin.ts() ])([ plugin.terser() ])
// => [ commonjs(), ts() ]
```

If you run `rollup -c --environment prod` it results in:

<!-- prettier-ignore -->
```ts
import { env } from '@brixtol/rollup-utils'


env.if('dev')([ plugin.commonjs(), plugin.ts() ])([ plugin.terser() ])
// => [ commonjs(), ts(), terser() ]
```

The value supplied must be a predefined value:

<!-- prettier-ignore -->
```ts
import { env } from '@brixtol/rollup-utils'

// VALID
env.if('dev')([])([])
env.if('prod')([])([])
env.if('watch')([])([])

// This will look for an env variable called "foo" and see if its value is "true" or "false"
env.if('foo')([])([])
```

Passing a `boolean` evaluation opposed to `dev`, `prod` or `watch` value:

<!-- prettier-ignore -->
```ts
import { env } from '@brixtol/rollup-utils'

// SAME AS PASSING "dev"
env.if(1 === 2)([])([])

// SAME AS PASSING "prod"
env.if(1 === 1)([])([])

```

#### `env.is`

Similar to the `env.if()` method the `env.is()` is a conditional executor. The `env.is()` method will either return the 2nd parameter when an `--environment` flag value is matched else it returns boolean `false` or `null` value (depending on the type passed). The 2nd parameter can be of a type `string`, `boolean` or rollup `plugin()`.

The `dev` is default, so running `rollup -c -w` results in:

<!-- prettier-ignore -->
```ts
import { env } from '@brixtol/rollup-utils'

// String
env.is('dev', 'hello world')    // => 'hello world'
env.is('watch', 'hello world')  // => 'hello world'

//
// This will return false as we called -c and -w
//
env.is('prod', 'hello world')   // => false

// Plugins
env.is('dev', terser())    // => terser()
env.is('watch', terser())  // => terser()

//
// This will return null as we called -c and -w
//
env.is('prod', terser())   // => null
```

If you run `rollup -c --environment prod` it results in:

<!-- prettier-ignore -->
```ts
import { env } from '@brixtol/rollup-utils'

// String
env.is('prod', 'hello world')   // => 'hello world'

//
// These will return false because we have an "--environment prod" flag
//
env.is('dev', 'hello world')    // => false
env.is('watch', 'hello world')  // => false

// Plugins
env.is('prod', terser())   // => terser()

//
// These will return null because we have an "--environment prod" flag
//
env.is('dev', terser())    // => null
env.is('watch', terser())  // => null
```

## `config.*`

The `config` export holds a parsed copy of different configuration files contained within your project. The methods give us a quick and effective way to read JSON files or get useful context about of current working directory.

#### `config.cwd`

Returns the current working directory path location (via `process.cwd()`).

#### `config.external`

Returns a `string[]` list of dependencies that can be passed to `rollup.external[]`.

#### `config.package`

Returns the `package.json` file contained in your project. This method will give you an jsdoc annotated typing reference so you can quickly access fields.

#### `config.tsconfig`

Returns the `tsconfig.json` file contained in your project. This method will give you an jsdoc annotated typing reference so you can quickly access fields. If no file is found it returns `null`.

#### `config.output{}`

Returns values contained within your projects `package.json` file that can passed to `rollup.output`. This is a sugar method, you can access also access these fields via `config.package`. If no value is found it returns `null`.

<!-- prettier-ignore -->
```ts
import { config } from '@brixtol/rollup-utils'

// returns the value of exports.require
config.output.cjs

// returns the value of exports.import
config.output.esm;

// returns the exports value (string or object)
config.exports;

// returns the value of the main field
config.main;

// returns the value of the module field
config.module;
```

## `date`

The `date` export returns UTC Time/Date in pretty format with suffixed number. Optionally accepts parameter value of UTC date string.

```ts
import { date } from '@brixtol/rollup-utils';

date(); // => 1st October 2021
```

## `path`

The `path` export is a resolver function that will resolve paths from the current working directory. This is just sugar for `path.resolve(process.cwd(), '/path/file.js')` but uses the cached `cwd` reference. Helpful if you need complete paths.

## `jsonmin`

The `jsonmin` export can be either used together with [rollup-plugin-copy](https://github.com/vladshcherbin/rollup-plugin-copy#readme) or process files separate from the rollup build. It will minify `.json` files. This export can also process JSON files using comments (`.jsonc`) and gracefully strip comments upon minification.

```ts
import { jsonmin } from '@brixtol/rollup-utils';

jsonmin(`/* comment */ { "field": "value" }`); // => {"field":"value"}
```

## `banner`

The `banner` export can be passed to `rollup.output.banner` which will return an inline Licence. The licenses will use fields contained within your `package.json` to inform upon project specifics in the licenses (like copyright, version, author etc).

In your `rollup.config.mjs` file:

```ts
import { banner } from '@brixtol/rollup-utils';

export default {
  input: 'src/file.js',
  output: {
    banner: banner('MIT' | 'PROPRIETARY' | 'REFER')
    // ...
  }
};
```

<details>
<summary>
PROPRIETARY
</summary>

The template literals represent fields in your `package.json` file.

```js
/**
 * !! THIS IS PROPRIETARY SOFTWARE !!
 *
 * @license
 *
 * ${basename(main)}
 *
 * Copyright © of ${owner} - All Rights Reserved.
 *
 * Unauthorized copying or modification of this file, via any medium is strictly
 * prohibited. Please refer to the LICENSE and/or ThirdPartyNotices.txt files
 * included in bundle.
 *
 * License:  ${license}
 * Package:  ${name}
 * Version:  ${version}
 * Updated:  ${date}
 *
 */
```

</details>

<details>
<summary>
MIT
</summary>

The template literals represent fields in your `package.json` file.

```js
/**
 * @license
 *
 * MIT License
 *
 * ${basename(main)}
 *
 * Copyright © ${owner}
 *
 * Package:  ${name}
 * Version:  ${version}
 * Updated:  ${date}
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
```

</details>

<details>
<summary>
REFER
</summary>

This option will require you to include the `LICENSE` as the generated text will refer to such a file. The template literals represent fields in your `package.json` file.

```js
/**
 * @license
 *
 * ${basename(main)}
 *
 * Copyright © ${owner}
 *
 * License:  ${license}
 * Package:  ${name}
 * Version:  ${version}
 * Updated:  ${date}
 *
 * Please refer to the LICENSE and/or ThirdPartyNotices.txt files included in bundle.
 *
 */
```

</details>

## Contributing

This package licensed under MIT but it exists as part of a monorepo that is closed source. Contributions, issues and/or feedback is welcome!

---

We [♡](https://www.brixtoltextiles.com/discount/4D3V3L0P3RS]) open source!


changelog.md
## _CHANGELOG_

### v1.2.1 | 31/03/2022

- Fixed type export on `config.alias`
- Enabled `object` to be passed to `alias`
- Remove a couple of dev dependencies

### v1.2.0 | 13/11/2021

- Fix `is` environment variable conditional for `dev` mode.
- Allow conditional boolean be passed to `if` and `is` wrappers

### v1.1.2 | 13/10/2021

- rollback concat and use spread

### v1.1.1 | 13/10/2021

- post-publish script

### v1.1.0 | 13/10/2021

- Added time utility which returns a pretty date

### v1.0.0 | 07/10/2021

- Cleaned up file
- Provided missing types
- Improved export methods in declarations
- Update readme.md and inform upon all available utils.

### v0.4.0 | 09/09/2021

- Fixed missing types used for package.json
- Applied declares for various types used by the tsconfig util

### v0.3.0 | 23/08/2021

- Provided getter `vars` to `env{}` method which returns vars in a `.env` file at cwd
- Added shortcuts to `exports`, `main` and `module` fields
- Updated types
- Cleaned up package.json file

### v0.2.0 | 17/08/2021

- Overhauled utilities
- Support for consuming configuration JSON files like `tsconfig.json` and `package.json`
- Added `env.*` support utility
- Extended `banner` utility
- Converted to `mjs` build export type
- Provided detailed typings for provided utilities

### v0.1.0 | 25/09/2020

Initial Release


test/rollup-utils.test.mjs
import test from 'ava';
import { jsonmin, env, config, path, date } from '../package/index.mjs';
import { readFileSync } from 'fs';

function fakeplugin () {
  return {
    load () {
      return null;
    }
  };
}

test.serial('Config export package.json config', t => {

  t.is(config.package.name, '@brixtol/rollup-utils');

  t.pass();

});

test.serial('Config export tsconfig.json config', t => {

  t.is(config.tsconfig, null);

  t.pass();

});

test.serial('Config current working directory', t => {

  t.is(config.cwd, process.cwd());

  t.pass();

});

test.serial('External export returning dependencies as an array', t => {

  t.deepEqual(config.external, Object.keys(config.package.dependencies));

  t.pass();

});

test.serial('Config path resolver', t => {

  const file = process.cwd() + '/test/jsonmin.jsonc';

  t.is(path('test/jsonmin.jsonc'), file);
  t.is(path('/test/jsonmin.jsonc'), '/test/jsonmin.jsonc');

  t.pass();

});

test.serial('Environment flag in dev mod', t => {

  t.is(env.watch, false);
  t.is(env.prod, false);
  t.is(env.dev, true);

  t.pass();

});

test.serial('Environment flag in prod mod', t => {

  process.env.prod = 'true';

  t.is(env.prod, true);
  t.is(env.watch, false);
  t.is(env.dev, false);

  delete process.env.prod;

  t.pass();

});

test.serial('Environment conditionals in dev mode', t => {

  process.env.dev = 'true';

  const plugin = fakeplugin();

  t.is(env.is('dev', plugin), plugin);
  t.is(env.is('dev', 'valid'), 'valid');
  t.is(env.is('prod', 'valid'), false);
  t.is(env.is('prod', plugin), null);

  const plugin1 = function devplugin () {};
  const plugin2 = function prodplugin () {};

  t.deepEqual(env.if('dev')([ plugin1 ])([ plugin2 ]), [ plugin1 ]);

  delete process.env.prod;

  t.pass();

});

test.serial('Environment conditionals in prod mode', t => {

  const plugin = fakeplugin();

  // dev is default, these will pass
  t.is(env.is('dev', plugin), plugin);
  t.is(env.is('dev', 'valid'), 'valid');

  process.env.prod = 'true';

  // these will now not work as we swapped to prod
  t.is(env.is('dev', 'valid'), false);
  t.is(env.is('dev', plugin), null);

  t.is(env.is('prod', 'valid'), 'valid');
  t.is(env.is('prod', plugin), plugin);

  const plugin1 = function devplugin () {};
  const plugin2 = function prodplugin () {};

  t.deepEqual(env.if('dev')([ plugin1 ])([ plugin2 ]), [ plugin1, plugin2 ]);

  t.pass();

});

test.serial('Minify JSON file with comments', t => {

  const json = readFileSync('./test/jsonmin.jsonc').toString();
  const minified = jsonmin(json);

  t.is(minified, '{"id":100,"text":"Hello World","array":[{"boolean":true,"object":{"number":1000.1}}]}');

  t.pass();

});

test.serial('Pretty formatted date', t => {

  t.log(date());

  t.pass();

});


index.js
import jsonStrip from 'strip-json-comments'
import jsonMinify from 'jsonminify'
import { basename, resolve } from 'path'
import { statSync } from 'fs-extra'
import stripIndent from 'strip-indent'


/* -------------------------------------------- */
/* PATH RESOLVER                                */
/* -------------------------------------------- */

/**
 * Path Resolver Helper
 *
 * @param {string} url
 * @returns {string}
 */
export const path = url => resolve(process.cwd(), url)

/* -------------------------------------------- */
/*                 JSON MINIFIER                */
/* -------------------------------------------- */

/**
 * Minify JSON and strip JSONC files
 *
 * @param {string} content
 */
export const jsonmin = (content) => {
  const parsed = JSON.parse(jsonStrip(content))
  const minified = jsonMinify(JSON.stringify(parsed))

  return minified
}

/* -------------------------------------------- */
/*                    PLUGINS                   */
/* -------------------------------------------- */

/**
 * Plugins - Concats development and production plugins
 * based on process env variable.
 *
 * @param {array} devPlugins
 * @param {array} prodPlugins
 */
export const plugins = (devPlugins, prodPlugins) => {

  if (process.env.prod) return [ ...devPlugins, ...prodPlugins ]

  return devPlugins

}

/* -------------------------------------------- */
/*                    BANNER                    */
/* -------------------------------------------- */

/**
 * License banner applied to javascript files
 *
 * @param {object} package
 * @returns {string}
 */
export const banner = (
  { name, main, version, author, owner, created },
  license = 'PROPRIETARY'
) => {
  owner = owner || author

  const date = new Date(statSync(main).mtimeMs)
    .toISOString()
    .replace(/T/, ' ')
    .substr(0, 19)

  switch (license) {
    case 'PROPRIETARY':
      return stripIndent(
        `
      /**
      *
      * !! THIS IS PROPRIETARY SOFTWARE !!
      *
      * @license
      *
      * ${basename(main)}
      *
      * Copyright © of ${owner} - All Rights Reserved.
      *
      * Unauthorized copying or modification of this file, via any medium is strictly
      * prohibited. Please refer to the LICENSE.txt and/or ThirdPartyNotices.txt files
      * included in bundle.
      *
      * License:  ${license}
      * Package:  ${name}
      * Version:  ${version}
      * Updated:  ${date}
      * Created:  ${created}
      *
      */`
      )
    case 'MIT':
    case 'CC BY-NC-ND 4.0':
      return stripIndent(
        `
      /**
      *
      * @license
      *
      * ${basename(main)}
      *
      * Copyright © ${owner}
      *
      * License:  ${license}
      * Package:  ${name}
      * Version:  ${version}
      * Updated:  ${date}
      * Created:  ${created}
      *
      * Please refer to the LICENSE.txt and/or ThirdPartyNotices.txt files included in bundle.
      *
      */`
      )
  }
}


rollup.config.mjs
import { defineConfig as Rollup } from 'rollup';
import cjs from '@rollup/plugin-commonjs';
import copy from 'rollup-plugin-copy';
import del from 'rollup-plugin-delete';
import { terser } from 'rollup-plugin-terser';
import { config, env } from './src/index.mjs';

export default Rollup(
  {
    input: 'src/index.mjs',
    output: [
      {
        format: 'cjs',
        file: config.output.cjs,
        sourcemap: env.is('dev', 'inline'),
        exports: 'named',
        esModule: false,
        preferConst: true
      },
      {
        format: 'es',
        file: config.output.esm,
        sourcemap: env.is('dev', 'inline'),
        preferConst: true
      }
    ],
    external: [
      'chalk',
      'jsonminify',
      'strip-indent',
      'strip-json-comments',
      'path',
      'fs',
      'dotenv'
    ],
    plugins: [
      del(
        {
          verbose: true,
          runOnce: !env.prod,
          targets: 'package/*'
        }
      ),
      copy(
        {
          verbose: true,
          copyOnce: !env.prod,
          targets: [
            {
              src: 'src/types/*.ts',
              dest: 'package'
            }
          ]
        }
      ),
      cjs(),
      terser(
        {
          ecma: 2016,
          warnings: 'verbose',
          compress: { passes: 2 }
        }
      )
    ]
  }
);


src/types/index.d.ts
import { PackageJSON, Exports } from './package';
import { TSConfig } from './tsconfig';
import { Plugin } from 'rollup';
import { DotenvParseOutput } from 'dotenv';

type EnvVars = 'prod' | 'dev' | 'watch';

type Env = (
  /**
   * The `--envronment` flag value
   */
  condition: EnvVars
) => (
  initial: string | string[] | Plugin[] | boolean
) => (combined: string | string[] | Plugin[] | boolean) => [];

/**
 * Resolves a path to the current working directory.
 *
 * @example
 *
 * path('file.js') // => 'User/dir/project/file.js'
 */
export declare const path: (uri: string) => string;

/**
 * Minify JSON and/or JSONC files.
 */
export declare const jsonmin: (content: string) => string;

/**
 * Get Time
 *
 * Returns UTC Time/Date or pretty format with suffixed number.
 * Optionally provide a predefined UTC date string.
 *
 * @example
 * // Without date or suffix params
 * 'Mon, 04 Oct 2021 02:53:57 GMT'
 *
 * // With date (includes suffix)
 * '4th October 2021'
 *
 */
export declare const date: (utc?: string | Date) => string;

/**
 * Banner - Applies a licence banner to output bundles generated
 * by rollup for proprietary software of the project
 */
export declare const banner: (
  license: 'MIT' | 'PROPRIETARY' | 'CC BY-NC-ND 4.0'
) => string;

/**
 * Validates env variable and executes condition based on result.
 */
export declare const env: {
  /**
   * Returns the parsed contents of a `.env` file
   */
  get vars(): DotenvParseOutput;
  /**
   * Sugar getter for `process.env.prod` flag, returning `true`
   * when in prod mode else `false`
   */
  get prod(): boolean;
  /**
   * Sugar getter for `process.env.prod` flag, returning `true`
   * when in dev mode else `false`
   */
  get dev(): boolean;
  /**
   * Sugar getter for `process.env.ROLLUP_WATCH` flag, returning `true`
   * when in watch mode else `false`
   */
  get watch(): boolean;
  /**
   * Tests a predefined condition environment variable. When true,
   * the `returns` value is returned, else `false`
   */
  is?: (
    condition: EnvVars,
    returns: boolean | string | Plugin | string[]
  ) => boolean | string | Plugin | string[];
  /**
   * Tests a condition for execution. It will concatenates returning array types.
   * by spreading. when condition equates to `false` combined is returned.
   *
   *
   * Plugins Example:
   *
   * ```js
   * // dev === true  > Returns initial array
   * env.if('dev')([plugin1()])([ plugin2() ]) => [ plugin1() ]
   *
   * // dev === false > Returns combined array
   * env.if('dev')([ plugin1() ])([ plugin2() ]) => [ plugin1(), plugin2() ]
   * ```
   *
   * String to Array Example:
   *
   * ```js
   * // dev === true  > Returns initial array
   * env.if('dev')(['foo','bar'])('baz') => ['foo','bar']
   *
   * // dev === false > Returns combined array
   * env.if('dev')(['foo','bar'])('baz') => ['foo','bar','baz']
   *
   * // dev === false > Returns combined array
   * env.if('dev')('foo')('bar') => ['foo', 'bar']
   * ```
   */
  if?: Env;
};

/**
 * Returns configuration of the current directory and package
 */
export declare const config: {
  /**
   * Returns current working directory path determined
   * by the value of `process.cwd()`
   */
  get cwd(): string;
  /**
   * JSON `package.json`
   */
  get package(): PackageJSON;
  /**
   * TypeScript Configuration File
   */
  get tsconfig(): TSConfig;
  /**
   * Returns `package.json` dependencies as keys
   */
  get external(): string[];
  /**
   * Returns the Current Working Directory relative
   * resolved path
   */
  path: (path: string) => string;
  /**
   * Return Alias mapping of paths, used in-conjunction
   * with `@rollup/plugin-alias` and passed to the `entries` field.
   *
   * @param {string[]} ids List of paths to alias
   * @param {string} [src] Directory to start mapping from
   */
  alias: (
    ids: string[] | { [path: string]: string[] },
    src?: string
  ) => Array<{
    find: string;
    replacement: string;
  }>;
  output: {
    /**
     * Returns the value of `exports.require` in the
     * projects `package.json` file.
     */
    get cjs(): string | null;
    /**
     * Returns the value of `exports.import` in the
     * projects `package.json` file.
     */
    get esm(): string | null;
    /**
     * Returns the `exports` value in the
     * projects `package.json` file.
     */
    get exports(): string | Exports | null;
    /**
     * Returns the value of the `main` field in the
     * projects `package.json` file.
     */
    get main(): string | null;
    /**
     * Returns the value of the `module` field in the
     * projects `package.json` file.
     */
    get module(): string | null;
  };
};


src/index.mjs
import jsonStrip from 'strip-json-comments';
import jsonMinify from 'jsonminify';
import { basename, join, resolve } from 'path';
import { readFileSync, existsSync } from 'fs';
import stripIndent from 'strip-indent';
import chalk from 'chalk';
import dotenv from 'dotenv';

/**
 * @type {import('./types/index')['env']}
 */
export const env = {
  get vars () {

    return dotenv.config();

  },
  get dev () {

    if (!this.prod) {
      process.env.dev = 'true';
      return true;
    } else {
      process.env.dev = 'false';
    }

    return false;

  },
  get prod () {

    if (process.env.prod === 'true') {
      process.env.dev = 'false';
      return true;
    };

    return false;

  },
  get watch () {

    return process.env.ROLLUP_WATCH === 'true';

  },
  is (condition, returns) {

    let environment;

    if (typeof condition === 'boolean') {
      environment = condition;
    } else {
      switch (condition) {
        case 'dev': environment = this.dev; break;
        case 'prod': environment = this.prod; break;
        case 'watch': environment = this.watch; break;
        default: environment = process.env[condition] === 'true';
      }
    }
    if (environment) return returns;

    return (
      typeof returns === 'string' ||
      typeof returns === 'number' ||
      typeof returns === 'boolean'
    ) ? false : null;

  },
  if (condition) {

    let environment;

    if (typeof condition === 'boolean') {

      environment = condition;

    } else {
      switch (condition) {
        case 'dev': environment = this.dev; break;
        case 'prod': environment = this.prod; break;
        case 'watch': environment = this.watch; break;
        default: environment = process.env[condition] === 'true';
      }
    }

    return initial => combined => {

      if (environment) return initial;

      const arrInitial = Array.isArray(initial);
      const arrCombined = Array.isArray(combined);

      if (arrInitial && arrCombined) return [ ...initial, ...combined ];

      const strInitial = typeof initial === 'string';
      const strCombined = typeof combined === 'string';

      if (arrInitial && strCombined) return [ ...initial, combined ];
      if (strInitial && arrCombined) return [ initial, ...combined ];
      if (strInitial && strCombined) return [ initial, combined ];

      const fnInitial = typeof initial === 'function';
      const fnCombined = typeof combined === 'function';

      if (arrInitial && fnCombined) return [ ...initial, combined ];
      if (fnInitial && arrCombined) return [ initial, ...combined ];
      if (fnInitial && fnCombined) return [ initial, combined ];

      return combined;

    };
  }
};

/**
 * @type {import('./types/index')['date']}
 */
export const date = (utc) => {

  if (!utc) utc = new Date();

  const time = new Date(utc).toLocaleDateString('en-gb', {
    year: 'numeric'
    , month: 'long'
    , day: 'numeric'
  });

  return time.replace(/([0-9]+)/, (n) => {

    const day = Number(n);

    return n + (
      [
        ''
        , 'st'
        , 'nd'
        , 'rd'
      ][day / 10 % 10 ^ 1 && day % 10] || 'th'
    );

  });

};

/**
 * @type {import('./types/index')['config']}
 */
export const config = (() => {

  const cwd = process.cwd();
  const file = readFileSync(join(cwd, 'package.json')).toString();
  const pkg = JSON.parse(jsonStrip(file));
  const path = join(cwd, 'tsconfig.json');

  return {
    get cwd () {

      return cwd;

    },
    get package () {

      return pkg;

    },
    get external () {

      return Object.keys(pkg.dependencies);

    },
    get tsconfig () {

      if (existsSync(path)) {
        const json = readFileSync(path).toString();
        const tsconfig = JSON.parse(jsonStrip(json));
        return tsconfig;
      }

      return null;

    },
    path: (uri) => {

      return resolve(cwd, uri);

    },
    alias: (ids, src = 'src') => {

      if (Array.isArray(ids)) {

        return ids.map(
          find => ({
            find: new RegExp(find),
            replacement: resolve(cwd, `${src}/${find}`)
          })
        );
      }

      if (typeof ids === 'object') {
        return Object.entries(ids).map(
          ([ find, [ value ] ]) => ({
            find,
            replacement: resolve(cwd, value)
          })
        );
      }
    },
    output: {
      get cjs () { return pkg.exports.require || null; },
      get esm () { return pkg.exports.import || null; },
      get exports () { return pkg.exports || null; },
      get main () { return pkg.main || null; },
      get module () { return pkg.module || null; }
    }
  };
})();

/**
 * @type {import('./types/index')['path']}
 */
export const path = (url) => resolve(process.cwd(), url);

/**
 * @type {import('./types/index')['jsonmin']}
 */
export const jsonmin = (content) => {

  if (content.length === 0) {
    return console.log(chalk`{italic  JSON file is empty, skipping}`);
  }

  try {
    const parsed = JSON.parse(jsonStrip(content));
    const minified = jsonMinify(JSON.stringify(parsed));

    return minified;
  } catch (e) {
    throw new Error(e);
  }

};

/**
 * @type {import('./types/index')['banner']}
 */
export const banner = (license = 'PROPRIETARY') => {

  const { name, main, version, author } = config.package;
  const owner = author;
  const date = new Date().toISOString().replace(/T/, ' ').substr(0, 19);

  switch (license) {
    case 'PROPRIETARY':
      return stripIndent(
        `
      /**
       * !! THIS IS PROPRIETARY SOFTWARE !!
       *
       * @license
       *
       * ${basename(main)}
       *
       * Copyright © of ${owner} - All Rights Reserved.
       *
       * Unauthorized copying or modification of this file, via any medium is strictly
       * prohibited. Please refer to the LICENSE and/or ThirdPartyNotices.txt files
       * included in bundle.
       *
       * License:  ${license}
       * Package:  ${name}
       * Version:  ${version}
       * Updated:  ${date}
       *
       */`
      );
    case 'MIT':
      return stripIndent(
        `
      /**
       * @license
       *
       * MIT License
       *
       * ${basename(main)}
       *
       * Copyright © ${owner}
       *
       * Package:  ${name}
       * Version:  ${version}
       * Updated:  ${date}
       *
       * Permission is hereby granted, free of charge, to any person obtaining a copy
       * of this software and associated documentation files (the "Software"), to deal
       * in the Software without restriction, including without limitation the rights
       * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
       * copies of the Software, and to permit persons to whom the Software is
       * furnished to do so, subject to the following conditions:
       *
       * The above copyright notice and this permission notice shall be included in all
       * copies or substantial portions of the Software.
       *
       * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
       * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
       * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
       * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
       * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
       * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
       * SOFTWARE.
       */`
      );
    case 'CC BY-NC-ND 4.0':
    case 'REFER':

      return stripIndent(
        `
      /**
       * @license
       *
       * ${basename(main)}
       *
       * Copyright © ${owner}
       *
       * License:  ${license}
       * Package:  ${name}
       * Version:  ${version}
       * Updated:  ${date}
       *
       * Please refer to the LICENSE.txt and/or ThirdPartyNotices.txt files included in bundle.
       *
       */`
      );
  }
};
